### DISCUSSION TITLE
Sales Channels & Apps Restrictions?

### DISCUSSION THREAD

**Discussion by bahamagician** [gatsbyjs/gatsby#32096] [Ideas / Feature Requests]
Previously, only products marked for inclusion in a particular Sales Channel or App was included in the products returned from the shopify api.  However, now I'm getting all of the products in the store regardless of whether they're meant to be included in the App that I've used to set up the API key.  Is there a way to exclude those products from the results returned?

**Comment by sslotsky** [comment]
Thanks for the issue. I'm not familiar with the concept of sales channels. Looking at the Admin API docs, it's possible to filter orders by sales channel, but I don't see any indication that other object types can be filtered this way. Do you happen to know where I can learn more about how sales channels work?

**Comment by div-cowboy** [comment]
@sslotsky Sales Channels are how Shopify controls what products show up on the Online Store, Shopify POS, Social Media (ads), and any Private App integrations (like this plugin). You can set availability on any product/collection, screenshots attached, and more insight [here](https://help.shopify.com/en/manual/online-sales-channels) 

[Screenshots](https://imgur.com/a/UJ0zdL0)

If I'm not mistaken, I believe that @bahamagician is pulling products that are not available on his Private App integration

**Comment by sslotsky** [comment]
Thanks for the info @iamkratos that was really helpful. It looks like the Admin API makes this available through publication data, but that data is only available for limited access scopes:

> read_publications access scope. This scope is currently available only to private apps installed on Shopify Plus stores.

We'll have to think on how to do this correctly. It will almost certainly involve opting in via `pluginOptions`. At that point we could either add an operation to fetch publication data and link the publications to the product nodes, or we could figure out how to incorporate `publichedOnCurrentPublication` into the products query. https://shopify.dev/docs/admin-api/graphql/reference/products-and-collections/publishable

**Comment by bahamagician** [comment]
Hey,

@iamkratos is correct.  Products not marked for inclusion in a particular sales channel are being included in the results pulled. 
 See the screenshots below which indicate where the feature is in the Shopify admin.  

![image](https://user-images.githubusercontent.com/13170078/113883034-cc5fa280-978b-11eb-85c9-9895c37e267d.png)

![image](https://user-images.githubusercontent.com/13170078/113887182-47768800-978f-11eb-9373-fe48fe05e765.png)

**Comment by lamellama** [comment]
This is causing problems adding items to cart because they should not be visible on the store

**Comment by sslotsky** [comment]
While we're still working this out, can folks here who are experiencing the issue let me know if my previous comment looks like it would be helpful for you? 

https://github.com/gatsbyjs/gatsby-source-shopify/issues/118#issuecomment-814900262

**Comment by lamellama** [comment]
publishedOnCurrentPublication (Boolean!)
Check to see whether the resource is published to the calling app's publication.

Sounds like what I'm expecting. I am surprised this is not the default and the API has access outside the channel

**Comment by sslotsky** [comment]
It's the admin API that we're using and I think the intention is that you can have an accurate reflection of what's really in the database. It makes sense why they would exclude these products from the Storefront API.

**Comment by sslotsky** [comment]
@lamellama follow up questions

1. In addition to supporting this field `publishedOnCurrentPublication` on both products and collections, is it also helpful to source [the list of all publications](https://shopify.dev/docs/admin-api/graphql/reference/apps/publicationconnection)?
2. We will have to support this optionally via pluginOptions, but if the answer to the above ☝🏼 is yes for at least some users, I'm wondering how many plugin options we need. Some folks might want the field without the list.

**Comment by sslotsky** [comment]
Here's what I'm thinking. We can add a new pluginOption called e.g. `privilegedAccessScopes`. If `read_publications` is included, then we will add `publishedOnCurrentPublication` to the product and collections queries.

Then if publication data is also desired, `publications` could be added to the `shopifyConnections` array in plugin options, which will only be considered valid if `read_publications` has been included in `privilegedAccessScopes`.

**Comment by sslotsky** [comment]
Actually I'm partly wrong about this! `publishedOnCurrentPublication` does not require privileged access. It _does_ requre the `read_product_listings` permission, but that's not limited to Shopify Plus stores so we can simply document this requirement.

So this really only leaves the question of whether or not you also need publication data to be sourced as well. If that's not useful to anyone, we can skip it and ship the new field that will hopefully help you solve your problem.

**Comment by lamellama** [comment]
I have currently modified the plugin to add publishedOnCurrentPublication to product and collection queries and then filtered my createPages accordingly which achieves what I need.

This is the only functionality I need but I am wondering if there is a more efficient way to avoid querying the extra data althogether.

**Comment by sslotsky** [comment]
> I am wondering if there is a more efficient way to avoid querying the extra data althogether.

Taking yet another look at the docs and I see one thing that could work. There is a filter key called `publishable_status` that you can feed to the products query. But so far I'm unable to find what the accepted values are for that filter or a description of what it does. If you find anything out here, by all means please let us know!

**Comment by sslotsky** [comment]
I've [inquired about this](https://community.shopify.com/c/Shopify-APIs-SDKs/Can-I-filter-Admin-API-products-by-sales-channel/m-p/1146761#M65863) with Shopify so hopefully we'll have an answer soon.

**Comment by div-cowboy** [comment]
@sslotsky Did some digging this weekend, and found [this](https://community.shopify.com/c/Shopify-APIs-SDKs/Product-published-status-vs-status/m-p/981882#M59819)

Could this provide the logic we need to resolve this issue?

**Comment by sslotsky** [comment]
It just might, thank you @iamkratos !!

**Comment by sslotsky** [comment]
Ok. let's make sure we get this right. According to that issue:

> Using the values published, unpublished, or any, the channel used is the online store.

So, is the online store the channel we want? Or do folks here need the ability to specify a different channel?

This morning I updated gatsbyjs/gatsby-source-shopify#135 to add this filter to both products and collections queries. I'd love for folks here to give that a try if you're able and let me know it solves your problem.

**Comment by waywardm** [comment]
When you create a private app to connect to Shopify such as 'Gatsby' that is then listed as an available sales channel. So it would make sense for the default to be online store with the option to have the 'Gatsby' Channel as the main preference. One problem with the shopify system is that when a new sales channel is created, all products are not enabled on that channel by default. So for a novice user, it would be best to have the default set to the online store but the ability to override with inputting your preferred sales channel in the config.

**Comment by div-cowboy** [comment]
@sslotsky No, we don't want the Online Store channel (the channel is typically used for product availability in a standard Shopify theme). When I create a new Private App in Shopify, it adds a new channel to all the product pages that shares the same name as the app. In the [screenshots attached](https://imgur.com/a/SXt84lr), I would want the channel to be the V8 (WMG)

**Comment by sslotsky** [comment]
Thanks y'all this is super duper helpful

**Comment by sslotsky** [comment]
Ok, I believe this is the solution being requested:

```ts
    const publishedStatus = this.pluginOptions.salesChannel
      ? encodeURIComponent(`${this.pluginOptions.salesChannel}=visible`)
      : `published`;

    const filters = [`status:active`, `published_status:${publishedStatus}`];
```

If you supply `salesChannel` via `pluginOptions`, you'll get `published_status:<salesChannel>=visible`. Otherwise, you'll get `published_status:published` which should default to the online store channel.

**Comment by div-cowboy** [comment]
@sslotsky looks slick 😎

**Comment by sslotsky** [comment]
One more question. Given the solution we've agreed on here, is it useful in any way to also include the following fields:

```
availablePublicationCount
publicationCount
publishedOnCurrentPublication
```

We can add them if they are useful, but there's some cost because they all require certain permissions.

**Comment by div-cowboy** [comment]
@sslotsky I don't think we need them. I never used these on the previous gatsby-source-shopify, and I'm not sure how I could make use of them in my builds.

**Comment by sslotsky** [comment]
Published in the newest `@rc`

**Comment by div-cowboy** [comment]
@sslotsky unfortunately, I think we have to re-open this. Although Shopify displays the Private App integration as a Sales Channel, it doesn't consider it one in the API. I've tried configuring my store use the new method but it wouldn't work.

**Comment by waywardm** [comment]
I tried the other day too stating a salesChannel and it didn’t filter them out.

Regards

Paul Kidd


Thunderfix
________________________________

**Comment by sslotsky** [comment]
> Although Shopify displays the Private App integration as a Sales Channel, it doesn't consider it one in the API. 

@iamkratos did you try that with their API directly? If it's true that their API does not consider the private app to be a sales channel, then I'm not sure how y'all want this to behave.

**Comment by waywardm** [comment]
![shopify](https://user-images.githubusercontent.com/12498554/117881161-d2700400-b2a0-11eb-8caf-aec564cdfbf7.PNG)

maybe this is the reason

taken from

https://shopify.dev/docs/admin-api/graphql/reference/products-and-collections/product#resourcepublications-2021-04

**Comment by lamellama** [comment]
seems sales channel scope is only for public apps: [https://community.shopify.com/c/Shopify-APIs-SDKs/Private-App-as-a-sales-channel/td-p/418433](https://community.shopify.com/c/Shopify-APIs-SDKs/Private-App-as-a-sales-channel/td-p/418433)

**Comment by fnebenfuehr** [comment]
Would be cool to get this working. I currently use a dirty graphql filter as workaround: 

```
allShopifyProduct(
      filter: { onlineStoreUrl: {regex: "/https?:\\/\\/(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)/"} }
      limit: 4
    ) {
  edges {
    node {
      id
      title
      }
    }
  }
```

**Comment by jackonawalk** [comment]
Here's a couple updates on sales channel support with `gatsby-source-shopify`

* Private apps are now able to function as a sales channel. This was a recent update from Shopify within the past couple weeks
* [Gatsby Cloud](https://apps.shopify.com/gatsby-storefront?surface_detail=gatsby&surface_inter_position=1&surface_intra_position=3&surface_type=search) is now available on Shopify's App Store which enables a "Gatsby Cloud" sales channel (no more private app configuration needed!). This works for new sites on Gatsby Cloud.

**Reply by div-cowboy** [reply]
@jacksellwood Thanks for looking into this Jack. I just updated all my plugins, and have tried:

- Sourcing products without a sales channel in my config (it's still pulling products that are only available on POS, not the Online Store channel)
- Adding sales channel as both regular case, and camel case to my gatsby config, but it's still not pulling the right product set (still pulling everything that is Active on any sales channel)

I'm trying the Gastby Cloud public app now, but I thought I'd mention these ^

**Reply by fnebenfuehr** [reply]
I can confirm. Having the same issue. Any updates on this?

**Reply by fnebenfuehr** [reply]
It turns out using the name of the sales channel doesn't work. Using the ID of the sales channel seems to work. https://github.com/gatsbyjs/gatsby/pull/32674#issue-959015709

The ID of the sales channel is displayed in the URL when filtering orders for sales channel (https://community.shopify.com/c/shopify-apis-and-sdks/how-to-get-id-for-my-shopify-app/m-p/527761/highlight/true#M34613)

**Reply by div-cowboy** [reply]
@jacksellwood has there been any progress on this issue? Gatsby does not update the products on my site when I update my product sales channel (it should remove the product, but the cached pdp page is still there)
